### PR TITLE
Make Asyncify work with wasm64

### PR DESCRIPTION
**Pull request by sbc100**
The emscripten side is a little tricky but I've got some tests passing. Currently blocked on:
https://github.com/emscripten-core/emscripten/issues/17969

### CODE REVIEW

File: src/passes/Asyncify.cpp [hunk 1]
```diff
@@ -795,30 +802,35 @@ static bool doesCall(Expression* curr) {
...
+                    makeGlobalGet(ASYNCIFY_DATA, pointerType),
+                    pointerType,
                     wasm.memories[0]->name);
   }
 
   Expression* makeIncStackPos(int32_t by) {
     if (by == 0) {
       return makeNop();
     }
+    Literal literal = is64 ? Literal(int64_t{by}) : Literal(int32_t{by});
```
Comment (kripken) [inline src/passes/Asyncify.cpp:826]:
```suggestion
    auto literal = Literal::makeFromInt64(by, pointerType);
```
I think with this we can avoid the `bool is64` on the class.

File: src/passes/Asyncify.cpp [hunk 2]
```diff
@@ -795,30 +802,35 @@ static bool doesCall(Expression* curr) {
...
     return makeStore(
       4,
       int32_t(DataOffset::BStackPos),
       4,
-      makeGlobalGet(ASYNCIFY_DATA, Type::i32),
-      makeBinary(AddInt32, makeGetStackPos(), makeConst(Literal(by))),
-      Type::i32,
+      makeGlobalGet(ASYNCIFY_DATA, pointerType),
+      makeBinary(is64 ? AddInt64 : AddInt32, makeGetStackPos(), makeConst(literal)),
```
Comment (kripken) [inline src/passes/Asyncify.cpp:832]:
```suggestion
      makeBinary(Abstract::getBinary(pointerType, Abstract::Add), makeGetStackPos(), makeConst(literal)),
```

File: src/passes/Asyncify.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/Asyncify.cpp:826]: (I mean with changes like this and the one below - then all we need is `pointerType`)

File: src/passes/Asyncify.cpp [hunk 3]
```diff
@@ -795,31 +802,36 @@ static bool doesCall(Expression* curr) {
 class AsyncifyBuilder : public Builder {
 public:
   Module& wasm;
+  Type pointerType;
 
-  AsyncifyBuilder(Module& wasm) : Builder(wasm), wasm(wasm) {}
+  AsyncifyBuilder(Module& wasm)
+    : Builder(wasm), wasm(wasm),
+      pointerType(wasm.memories[0]->is64() ? Type::i64 : Type::i32) {}
```
Comment (kripken) [inline src/passes/Asyncify.cpp:809]:
```suggestion
      pointerType(wasm.memories[0]->indexType) {}
```

File: src/passes/Asyncify.cpp [hunk 4]
```diff
@@ -1497,6 +1510,8 @@ struct Asyncify : public Pass {
   void run(Module* module) override {
     auto& options = getPassOptions();
     bool optimize = options.optimizeLevel > 0;
+    is64 = module->memories.size() && module->memories[0]->is64();
+    pointerType = is64 ? Type::i64 : Type::i32;
```
Comment (kripken) [inline src/passes/Asyncify.cpp:1514]:
```suggestion
    pointerType = module->memories[0]->indexType;
```

File: src/passes/Asyncify.cpp [hunk 5]
```diff
@@ -795,31 +802,36 @@ static bool doesCall(Expression* curr) {
...
   Module& wasm;
+  Type pointerType;
 
-  AsyncifyBuilder(Module& wasm) : Builder(wasm), wasm(wasm) {}
+  AsyncifyBuilder(Module& wasm)
+    : Builder(wasm), wasm(wasm),
+      pointerType(wasm.memories[0]->is64() ? Type::i64 : Type::i32) {}
 
   Expression* makeGetStackPos() {
     return makeLoad(4,
```
Comment (kripken) [inline src/passes/Asyncify.cpp:812]: Doesn't the load size need to be 8 now? This is loading a pointer, which is 8 bytes.

File: src/passes/Asyncify.cpp (hunk 4 above)
Comment (sbc100) [reply src/passes/Asyncify.cpp:1514]: IIRC the problem here is that module->memories might be empty.   I guess asyncify should probably error out in that case?    But that seems like a bigger change.

Comment (kripken) [reply src/passes/Asyncify.cpp:1514]: Oh, right, sorry, my suggested diff here isn't right then. Still, using `indexType` might simplify things.

File: src/passes/Asyncify.cpp (hunk 5 above)
Comment (sbc100) [reply src/passes/Asyncify.cpp:812]: Yes, you are right, updated here and elsewhere.

**Review by kripken — APPROVED** [review]

**Pull request merged**